@here/harp-mapview/lib/MapView.ts

```diff
@@ -2613,6 +2655,133 @@ export class MapView extends THREE.EventDispatcher {
+     * @param result Result to place calculation.
+     */
+    private ndcToView(vector: Vector3Like, result: THREE.Vector3): THREE.Vector3 {
+        result
+            .set(vector.x, vector.y, vector.z)
+            .applyMatrix4(this.camera.projectionMatrixInverse)
+            // Make sure to apply rotation, hence use the rte camera
+            .applyMatrix4(this.m_rteCamera.matrixWorld);
+        return result;
+    }
+
+    /**
+     * Transfer from view space to camera space.
+     * @param viewPos position in view space, result is stored here.
+     */
+    private viewToLightSpace(viewPos: THREE.Vector3, camera: THREE.Camera): THREE.Vector3 {
+        return viewPos.applyMatrix4(camera.matrixWorldInverse);
+    }
+
+    /**
+     * Update the directional light camera. Note, this requires the cameras to first be updated.
+     */
+    private updateLights() {
+        // TODO: HARP-9479 Globe doesn't support shadows.
+        if (
+            !this.shadowsEnabled ||
+            this.projection.type === ProjectionType.Spherical ||
+            this.m_createdLights === undefined ||
+            this.m_createdLights.length === 0
+        ) {
+            return;
+        }
+
+        const points: Vector3Like[] = [
+            // near plane points
+            { x: -1, y: -1, z: -1 },
+            { x: 1, y: -1, z: -1 },
+            { x: -1, y: 1, z: -1 },
+            { x: 1, y: 1, z: -1 },
+
+            // far planes points
+            { x: -1, y: -1, z: 1 },
+            { x: 1, y: -1, z: 1 },
+            { x: -1, y: 1, z: 1 },
+            { x: 1, y: 1, z: 1 }
+        ];
+        const transformedPoints = points.map((p, i) => this.ndcToView(p, cache.frustumPoints[i]));
+
+        this.m_createdLights.forEach(element => {
+            const directionalLight = element as THREE.DirectionalLight;
+            if (directionalLight.isDirectionalLight === true) {
+                const pointsInLightSpace = transformedPoints.map(p =>
+                    this.viewToLightSpace(p, directionalLight.shadow.camera)
+                );
+
+                const box = new THREE.Box3();
+                pointsInLightSpace.forEach(point => {
+                    box.expandByPoint(point);
+                });
+                const camera = directionalLight.shadow.camera;
+                camera.left = box.min.x;
+                camera.right = box.max.x;
+                camera.top = box.max.y;
+                camera.bottom = box.min.y;
+                camera.near = -box.max.z;
+                camera.far = -box.min.z;
+                camera.updateProjectionMatrix();
+
+                const lightDirection = cache.vector3[0];
+                lightDirection.copy(directionalLight.target.position);
+                lightDirection.sub(directionalLight.position);
+                lightDirection.normalize();
+
+                const target = MapViewUtils.getWorldTargetFromCamera(this.camera, this.projection);
```

Comment:
Shadows should also work if no target point exists. I recently merged my PR #1365 that ensures that `MapView.m_targetWorld` always exists, so maybe you should use that

Reply:
I see: `@deprecated This function is for internal use only and will be removed in the future.`, which doesn’t unfortunately give me any idea how to upgrade, some help would be greatly apprecated. 🙂